crates/nu-parser/src/type_check.rs

```diff
@@ -7,16 +7,18 @@ use nu_protocol::{
 pub fn type_compatible(lhs: &Type, rhs: &Type) -> bool {
     // Structural subtyping
     let is_compatible = |expected: &[(String, Type)], found: &[(String, Type)]| {
-        // the expected type is `any`
         if expected.is_empty() {
             true
-        } else if expected.len() != found.len() {
+        } else if expected.len() > found.len() {
             false
         } else {
-            expected
-                .iter()
-                .zip(found.iter())
-                .all(|(lhs, rhs)| lhs.0 == rhs.0 && type_compatible(&lhs.1, &rhs.1))
+            expected.iter().all(|(col_x, ty_x)| {
+                if let Some((_, ty_y)) = found.iter().find(|(col_y, _)| col_x == col_y) {
+                    type_compatible(ty_x, ty_y)
+                } else {
+                    false
+                }
+            })
```

Comment:
Just thinking out loud regarding worst case performance.

Given required type fields `r` and found type fields `f` this turns the operation from `O(r)` to `O(r log r)`
I was initially wondering if it makes sense to enforce our `Type:Record`/`Table` to be sorted. but this would require `O(n log n)` initally and makes some other things like displaying the type less nice.
So I think this is the price to pay for good structural subtyping, and shouldn't become too much of a practical issue as the enforced types are probably small (sorting every type of each large table record would probably worse in the long run)

Reply:
good observation, but as you said, the average table doesn't have very many columns

we could also sort `expected` (in that closure), and not touch the other stuff like display

Reply:
I think let's wait for this to pop up in a profile before we optimize. (admittedly sorting `expected` and then doing binary search sounds nice from theoretical runtime but is probably a negative in real-world memory behavior on small sizes)